pkg/controller/mysqlcluster/mysqlcluster_controller.go

```diff
@@ -109,58 +126,119 @@ func (r *ReconcileMysqlCluster) Reconcile(request reconcile.Request) (reconcile.
 	}
 
-	// TODO(user): Change this for the object type created by your controller
-	// Check if the Deployment already exists
-	found := &appsv1.Deployment{}
-	err = r.Get(context.TODO(), types.NamespacedName{Name: deploy.Name, Namespace: deploy.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		log.Printf("Creating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
-		err = r.Create(context.TODO(), deploy)
+	// TODO: set default on a copy cluster
+	cluster.SetDefaults(r.opt)
+
+	defer func() {
+		// TODO: update just status and not in a defer
+		err := r.Update(context.TODO(), cluster) // nolint
 		if err != nil {
-			return reconcile.Result{}, err
+			log.Error(err, "Failed to update cluster status!", "cluster", cluster)
 		}
-	} else if err != nil {
+	}()
+
+	// run the config syncers
+	configSyncers := []syncers.Interface{
+		mysqlcluster.NewConfigMapSyncer(cluster),
+		mysqlcluster.NewSecretSyncer(cluster),
+	}
+
+	err = r.sync(cluster, configSyncers)
+	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// TODO(user): Change this for the object type created by your controller
-	// Update the found object and write the result back if there are any changes
-	if !reflect.DeepEqual(deploy.Spec, found.Spec) {
-		found.Spec = deploy.Spec
-		log.Printf("Updating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
-		err = r.Update(context.TODO(), found)
+	// get config map and secret revision to annotate the statefulset
+	// TODO: update this, no need to get secret and config again, already are updated
+	// in existing object provided to CreateOrUpdate
+	configRev, secretRev, err := r.getConfigAndSecretRevs(cluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// run the syncers for services, pdb and statefulset
+	otherSyncers := []syncers.Interface{
+		mysqlcluster.NewHeadlessSVCSyncer(cluster),
+		mysqlcluster.NewMasterSVCSyncer(cluster),
+		mysqlcluster.NewHealthySVCSyncer(cluster),
+
+		mysqlcluster.NewStatefulSetSyncer(cluster, configRev, secretRev, r.opt),
+	}
+
+	if len(cluster.Spec.MinAvailable) != 0 {
+		otherSyncers = append(otherSyncers, mysqlcluster.NewPDBSyncer(cluster))
+	}
+
+	return reconcile.Result{}, r.sync(cluster, otherSyncers)
+}
+
+func (r *ReconcileMysqlCluster) sync(cluster *mysqlv1alpha1.MysqlCluster, syncers []syncers.Interface) error {
+	for _, s := range syncers {
+		existing := s.GetExistingObjectPlaceholder()
+
+		// set owner reference on objects that requires owner reference
+		syncFn := s.Sync
+		if s.ShouldHaveOwnerReference() {
+			syncFn = func(in runtime.Object) error {
+				// TODO: check this
+				metaIn := in.(metav1.Object)
+				if err := controllerutil.SetControllerReference(cluster, metaIn, r.scheme); err != nil {
+					return err
+				}
+				return s.Sync(in)
+			}
+		}
+
+		op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, existing, syncFn)
+		reason := getErrorEventReason(existing, err)
```

Comment:
`syncers.Interface` should have a method for this, like the one in presslabs/wordpress-operator and presslabs/dashboard have.

Reply:
Its mot very useful, because a generated string returned by `getErrorEventReason`, from object type, does as well as using a custom function in Interface. And using a custom function will finally end up in duplicated code.